### PR TITLE
Fix the autoattack wrapper

### LINKED ISSUE #79: AutoAttack does not run
```console
$ python -m mart experiment=CIFAR10_RobustBench trainer=gpu fit=false +trainer.limit_test_batches=1 +attack@model.modules.input_adv_test=classification_autoattack

[2023-02-16 11:35:14,513][mart.utils.utils][ERROR] -
Traceback (most recent call last):
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 92, in _call_target
    return _target_(*args, **kwargs)
TypeError: __init__() got an unexpected keyword argument 'adversary'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/weilinxu/coder/MART/mart/utils/utils.py", line 54, in wrap
    metric_dict, object_dict = task_func(cfg=cfg)
  File "/home/weilinxu/coder/MART/mart/tasks/lightning.py", line 39, in lightning
    model: LightningModule = hydra.utils.instantiate(cfg.model)
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 222, in instantiate
    return instantiate_node(
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 334, in instantiate_node
    value = instantiate_node(
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 349, in instantiate_node
    dict_items[key] = instantiate_node(
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 339, in instantiate_node
    return _call_target(_target_, partial, args, kwargs, full_key)
  File "/home/weilinxu/coder/MART/.venv/lib/python3.9/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 97, in _call_target
    raise InstantiationException(msg) from e
hydra.errors.InstantiationException: Error in call to target 'mart.attack.adversary_wrapper.NormalizedAdversaryAdapter':
TypeError("__init__() got an unexpected keyword argument 'adversary'")
full_key: model.modules.input_adv_test
[2023-02-16 11:35:14,514][mart.utils.utils][INFO] - Closing loggers...
Error executing job with overrides: ['experiment=CIFAR10_RobustBench', 'trainer=gpu', 'fit=false', '+trainer.limit_test_batches=1', '+attack@model.modules.input_adv_test=classification_autoattack']
Error in call to target 'mart.attack.adversary_wrapper.NormalizedAdversaryAdapter':
TypeError("__init__() got an unexpected keyword argument 'adversary'")
full_key: model.modules.input_adv_test

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

### PR DESCRIPTION
**Pull request by mzweilin**
# What does this PR do?

Fixes #79

Test run on GPU, attacking a pretrained model. 
```bash
python -m mart \
experiment=CIFAR10_RobustBench \
fit=false \
+attack@model.modules.input_adv_test=classification_autoattack \
trainer=ddp +trainer.limit_test_batches=1
```

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] `pytest tests/test_experiments.py::test_cifar10_cnn_autoattack_experiment`

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/utils/adapters.py [hunk 1]
```diff
@@ -31,6 +33,11 @@ def __call__(self, *args, **kwargs):
             args (Any): values to use in the callable method.
             kwargs (Any): keyword values to use in the callable method.
         """
+        # First make a partial to a function.
+        if isinstance(self.instance, functools.partial):
+            self.instance = self.instance(*args, **kwargs)
+            return self
+
```
Comment (dxoigmn) [inline mart/utils/adapters.py:40]: Why is this necessary?

**Review by dxoigmn — CHANGES_REQUESTED** [review]
Can you add a test? Ideally it would fail on the old code.

File: mart/utils/adapters.py [hunk 2]
```diff
@@ -5,34 +5,51 @@
...
         Args:
-            args (Any): values to use in the callable method.
-            kwargs (Any): keyword values to use in the callable method.
+            instance (object): instance to make callable.
         """
-        function = getattr(self.instance, self.redirecting_fn)
 
+        function = getattr(instance, self.redirecting_fn)
         assert callable(function)
+        return function
```
Comment (dxoigmn) [inline mart/utils/adapters.py:35]: I don't quite get why any of this code has to change?

File: mart/utils/adapters.py [hunk 3]
```diff
@@ -5,34 +5,51 @@
...
+
+
+class PartialAdapter:
```
Comment (dxoigmn) [inline mart/utils/adapters.py:38]: I would probably name this `PartialInstanceWrapper` or just `InstanceWrapper`.

File: mart/utils/adapters.py (hunk 2 above)
Comment (dxoigmn) [reply mart/utils/adapters.py:35]: Yeah, see if you can keep this code the same while only change the new class.

**Review by dxoigmn — CHANGES_REQUESTED** [review]
See comments.

File: mart/utils/adapters.py (hunk 2 above)
Comment (mzweilin) [reply mart/utils/adapters.py]: OK. I would hardcode `CallableAdapter`.

File: mart/utils/adapters.py (hunk 3 above)
Comment (mzweilin) [reply mart/utils/adapters.py]: Good idea. Renamed as `PartialInstanceWrapper`.

File: mart/utils/adapters.py [hunk 4]
```diff
@@ -36,3 +38,23 @@ def __call__(self, *args, **kwargs):
...
+            redirecting_fn (str): name of the function that will be invoked in the `__call__` method.
+        """
+        self.partial = partial
+        self.redirecting_fn = redirecting_fn
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Callable:
+        # Turn a partial to a class object.
+        instance = self.partial(*args, **kwargs)
+        # Make the object callable.
+        return CallableAdapter(instance, self.redirecting_fn)
```
Comment (dxoigmn) [inline mart/utils/adapters.py:60]: This just copies the functionality of CallableAdapter. Can you rewrite this so that you don't tightly couple `PartialInstanceWrapper` with `CallableAdapter`?

**Review by dxoigmn — CHANGES_REQUESTED** [review]
See comments. Also needs a test.

File: mart/utils/adapters.py (hunk 4 above)
Comment (mzweilin) [reply mart/utils/adapters.py]: Fixed.

File: mart/utils/adapters.py [hunk 5]
```diff
@@ -36,3 +38,23 @@ def __call__(self, *args, **kwargs):
...
+
+    def __init__(self, partial: Callable, adapter):
+        """
+
+        Args:
+            partial (Callable): A partial of a class object.
+            adapter (CallableAdapter): An adapter that creates the `__call__` method.
+        """
+        self.partial = partial
+        self.adapter = adapter
```
Comment (dxoigmn) [inline mart/utils/adapters.py:54]: I would name this `wrapper` given the name of the class. I also think the only type requirement on this object that it is Callable.

File: mart/utils/adapters.py (hunk 4 above)
Comment (dxoigmn) [reply mart/utils/adapters.py]: Looks good!

File: mart/utils/adapters.py (hunk 5 above)
Comment (mzweilin) [reply mart/utils/adapters.py]: Fixed.

**Review by dxoigmn — APPROVED** [review]
LGTM!
